### PR TITLE
Fixed #29132 -- Avoided connecting update_last_login() handler if User.last_login isn't a field.

### PR DESCRIPTION
**Pull request by marazmiki**
https://code.djangoproject.com/ticket/29132

### CODE REVIEW

**Comment by timgraham** [comment]
To fix this PR, you should cherry pick your commit to the master branch rather than branching off of stable/2.0.x and force push that new branch here.

**Comment by marazmiki** [comment]
Oh, got it

File: tests/auth_tests/test_signals.py [hunk 1]
```diff
@@ -97,10 +97,15 @@ def test_login_with_custom_user_without_last_login_field(self):
...
                 apps.get_app_config('auth').ready()
             self.assertEqual(signals.user_logged_in.receivers, [])
 
+            with self.settings(AUTH_USER_MODEL='auth_tests.UserWithDisabledLastLoginField'):
+                apps.get_app_config('auth').ready()
+            self.assertEqual(signals.user_logged_in.receivers, [])
+
             with self.settings(AUTH_USER_MODEL='auth.User'):
                 apps.get_app_config('auth').ready()
             self.assertEqual(len(signals.user_logged_in.receivers), 1)
```
Comment (carltongibson) [inline tests/auth_tests/test_signals.py]: These checks aren't quite independent. When this one is first it fails. But I guess it's OK.

Comment (marazmiki) [reply tests/auth_tests/test_signals.py]: Personally, I would not like to change an existing test, as it checking some fixed behavior. In my mind, it's different cases. If I've screwed up with a new condition, the first test will tell me about it.

Comment (carltongibson) [reply tests/auth_tests/test_signals.py]: No problem. I think it’s OK as is. (Adjusting it would be done separately anyway)

Comment (marazmiki) [reply tests/auth_tests/test_signals.py]: @carltongibson can you please check it?

Comment (carltongibson) [reply tests/auth_tests/test_signals.py]: Yep. Looks fine. I've marked it Ready for Checkin. Most likely Tim will have a look and merge it if he's happy.

File: django/contrib/auth/apps.py [hunk 2]
```diff
@@ -18,8 +19,10 @@ def ready(self):
             create_permissions,
             dispatch_uid="django.contrib.auth.management.create_permissions"
         )
-        if hasattr(get_user_model(), 'last_login'):
+        last_login_field = getattr(get_user_model(), 'last_login', None)
+        if isinstance(last_login_field, DeferredAttribute):
             from .models import update_last_login
             user_logged_in.connect(update_last_login, dispatch_uid='update_last_login')
+
```
Comment (felixxm) [inline django/contrib/auth/apps.py]: Please chop this unrelated blank line.

File: tests/auth_tests/test_signals.py [hunk 3]
```diff
@@ -97,10 +97,15 @@ def test_login_with_custom_user_without_last_login_field(self):
             signals.user_logged_in.receivers = []
             with self.assertRaises(FieldDoesNotExist):
                 MinimalUser._meta.get_field('last_login')
+
```
Comment (felixxm) [inline tests/auth_tests/test_signals.py]: Please chop this unrelated blank line.

File: django/contrib/auth/apps.py (hunk 2 above)
Comment (carltongibson) [reply django/contrib/auth/apps.py]: Hey @felixxm. I looked at both of these and thought them OK because they aided readability. (Just a small visual separator between the blocks). 

Is your objection that to the lines themselves, or that they're unrelated to the change per se? If the latter would a separate commit be better? 

(Thanks!)

Comment (timgraham) [reply django/contrib/auth/apps.py]: I wouldn't add the extra lines for both reasons. Not to say that mistakes aren't made, but generally if we merged a previous patch without them (at least recently when code style is more standardized than in the early days), then blank line changes probably aren't needed. Certainly putting unrelated blank line additions in a patch doesn't help keep things clean and I don't think the blank lines are so important that they deserve a commit.

Comment (carltongibson) [reply django/contrib/auth/apps.py]: Hmmm... Not sure I entirely agree. (Which is OK.)

In general it seems to leave no space for the little changes which over-time add up to more readable code: can't add to a change commit, because they're unrelated, can't add as a separate commit because (yes, you're absolutely right) individually they don't add up to much.

In particular the whitespacing in the (other) `test_signals` block isn't even consistent within the single test case — I thought the extra line there just such one of these insignificant cases that is better afterwards (since the spacing is then consistent).  

Anyhow, no a biggie here.

Comment (timgraham) [reply django/contrib/auth/apps.py]: The reason there's no whitespace after `with self.assertRaises(FieldDoesNotExist):` is because that's logically connected to the with statement that follows. I added `self.assertIsNone(UserWithDisabledLastLoginField().last_login)` to the new test which serves a similar purpose.

Comment (carltongibson) [reply django/contrib/auth/apps.py]: OK. Makes sense. Ta!
